dts: bindings: Update Nordic owned memory bindings

57300 opened the pull request:
This concerns both `nordic,owned-memory` and `nordic,owned-partitions`.

Introduce a property named `nordic,access`, which is meant to replace
the `owner-id` and `perm-*` properties. It allows for describing how
multiple domains should access a single memory region, possibly with
different permissions per owner, but without having to create more than
one DT node for this purpose.

This change is also motivated by updated memory protection requirements
on the nRF54H20, which mandate that a given memory region must only be
reserved by one domain, even if multiple domains can have access to it.
This restriction is now described in the binding itself.

57300 committed:
dts: bindings: Update Nordic owned memory bindings

This concerns both `nordic,owned-memory` and `nordic,owned-partitions`.

Introduce a property named `nordic,access`, which is meant to replace
the `owner-id` and `perm-*` properties. It allows for describing how
multiple domains should access a single memory region, possibly with
different permissions per owner, but without having to create more than
one DT node for this purpose.

This change is also motivated by updated memory protection requirements
on the nRF54H20, which mandate that a given memory region must only be
reserved by one domain, even if multiple domains can have access to it.
This restriction is now described in the binding itself.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
boards: nordic: Align with updated Nordic owned memory bindings

Apply the following changes to `nrf54h20dk` and `nrf9280pdk`:

* Convert `perm-*` properties to the newly introduced `nordic,access`,
  both in board files and tests.

* Redefine shared regions to specify multiple access owners per node,
  and ensure that each such region is reserved by one domain at a time.
  `cpuapp_cpurad_ram0x_region` is only enabled by Radiocore, while
  `cpuapp_cpucell_ram0x_region` is only enabled by Application core.

* Divide `shared_ram3x_region` so that each sub-region is owned by a
  different domain. Their addresses must be rounded down to fit the
  current UICR format.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

gmarull commented on include/zephyr/dt-bindings/reserved-memory/nordic-owned-memory.h:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @param ... List of permissions, such as READ, named after NRF_PERM_* definitions.
+ */
+#define NRF_PERM(...) (FOR_EACH_FIXED_ARG(PREPEND, (|), NRF_PERM_, __VA_ARGS__))
```
not a fan of macrology in DT, we could just tell users to use eg `NRF_PERM_READ | NRF_PERM_WRITE`.

gmarull commented on dts/bindings/reserved-memory/nordic,owned-memory.yaml:
```diff
@@ -8,17 +8,63 @@ description: |
+
   owner-id:
     type: int
+    deprecated: true
```
is it worth deprecating at this stage?

57300 replied:
Yes, maybe not for Zephyr but for downstream, deprecation is practical in the short-term.

gmarull commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-memory_map.dtsi:
```diff
@@ -10,9 +10,7 @@
-			perm-read;
-			perm-write;
-			perm-secure;
+			nordic,access = <NRF_OWNER_ID_APPLICATION NRF_PERM(READ, WRITE, SECURE)>;
```
true that macrology allows for _nicer_ and shorter content

gmarull commented on tests/arch/common/ramfunc/boards/nrf54h20dk_nrf54h20_cpuapp.overlay:
```diff
@@ -4,5 +4,5 @@
 
 &cpuapp_ram0x_region {
-	perm-execute;
+	nordic,access = <NRF_OWNER_ID_APPLICATION NRF_PERM(READ, WRITE, EXEC, SECURE)>;
```
suggestion, to avoid macrology, why don't we provide:

```c
#define NRF_PERM_R x
#define NRF_PERM_W y
...
#define NRF_PERM_RW NRF_PERM_R | NRF_PERM_W
...
```

so DT would be even less verbose, and, no macrology is involved.

57300 replied:
Sure. This would produce 31 definitions in total, but I don't mind it if it can be accepted.

gmarull replied:
I, personally, prefer that, it's one-time definition work, also autocomplete will work out of the box.

57300 replied:
Applied.

gmarull reviewed (dismissed):
lgtm, thanks!

hubertmis commented on dts/bindings/reserved-memory/nordic,owned-memory.yaml:
```diff
@@ -8,17 +8,64 @@ description: |
+    };
+
+  A single local domain can request a memory region to be reserved on behalf of
+  multiple access owners. A single memory region shall be reserved by at most
```
This rule would require a complex logic on the secure domain firmware. Let's discuss the following example planned in the memory map:

There are three concatenated RAM regions:

* Application's data
* Aplication - Radio shared IPC data
* Radio's data.

We have only two available MPC OVERRIDEs to describe these regions:

* Owner App, perm rw, adresses covering regions 1 and 2
* Owner Radio, perm rw, addresses covering regions 2 and 3

In your proposal build-time tools would need to verify that the three defined memory regions are concatenated in the memory. An algorithm would need to discover concatenated memory regions with the same owner and permissions and optimize number of allocated MPC OVERRIDEs.

Instead, I would suggest to make dts nodes representing memory configured in dts overrides like this:

```
app_rw_regions:
  nordic-access = application rw
  reg = <reg1start, size_of_regs_1and2>

radio_rw_regions:
  nordic-access = radio rw
  reg = <reg2start, size_of_regs_2and3>
```

Both nodes would includr sub-nodes describing nodes 1, 2; and 2, 3 accordingly.

This way mapping of the memory regions to MPC would be present explicilty in dts helping to design the memory layout properly and avoiding build time errors when a memory layout with implicit MPC mapping could not be translated to the available number of MPC entries.

masz-nordic commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -8,13 +8,19 @@
 #define NRF_DOMAIN_ID_GLOBALFAST  12
 #define NRF_DOMAIN_ID_GLOBALSLOW  13
 
+/* Owner IDs. Can be used to specify memory access. */
+#define NRF_OWNER_ID_NONE        0
+#define NRF_OWNER_ID_APPLICATION 2
+#define NRF_OWNER_ID_RADIOCORE   3
```
Those magic numbers should either be validated against values in MDK or generated from SVD.

carlescufi replied:
@57300 I think @gmarull has a script to generate from SVD that you could use here.

57300 replied:
@carlescufi That's true, but these values aren't found in SVD. I will add MDK validation in another PR.

carlescufi replied:
@57300 would you mind adding the MDK validation directly here?

57300 replied:
> @57300 would you mind adding the MDK validation directly here?

I'm still waiting for other dependency PRs to get merged before I rebase.
In the meantime, I opened a new PR for this (you'll see why): https://github.com/zephyrproject-rtos/zephyr/pull/72507

hubertmis reviewed (dismissed):
As discussed in https://github.com/nrfconnect/sdk-secdom/pull/688, a run-time merging algorithm will be implemented. The overlapping of memory regions between UICRs should be prohibited.

hubertmis commented:
@57300 , I understood this change should be merged. Do you plan to continue working on this?

57300 commented:
> @57300 , I understood this change should be merged. Do you plan to continue working on this?

Yes, I will reopen once internal dependencies are resolved.

rettichschnidi commented:
I think I'll need this for #78253.

57300 commented:
Reopened, rebased, and added new changes for `nrf9280pdk`.

hubertmis commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpurad.dts:
```diff
@@ -34,10 +34,6 @@
 	};
 };
 
-&shared_ram3x_region {
```
And this with `&cpurad_dma_region`?

57300 replied:
`cpurad_dma_region` is already enabled further down.

hubertmis replied:
Missed this, thanks!

57300 commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-memory_map.dtsi:
```diff
@@ -55,9 +51,7 @@
-			perm-read;
-			perm-write;
-			perm-secure;
+			nordic,access = <NRF_OWNER_ID_APPLICATION NRF_PERM_RWS>;
```
I'd say this is out of scope of this PR. I'm unsure what conditions need to be met before the `TODO` can be addressed.

hubertmis replied:
I think the `TODO` was blocked by this PR. But I'm fine to remove it in another.

57300 commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-memory_map.dtsi:
```diff
@@ -172,29 +163,27 @@
 		};
 
 		shared_ram3x_region: memory@2fc12000 {
```
I don't have a strong reason to keep it. There's a similar situation with the `shared_ram20_region` node, which also hasn't been `nordic,owned-memory` compatible for a while, so it could probably be flattened as well. The only thing is that I would want to rename `cpuapp_dma_region` to `cpuapp_dma_ram3x_region` etc. to be more consistent; this would create a bigger diff, since these nodes are referenced in several overlays.

hubertmis replied:
I agree with your proposal. I believe the additional nesting was caused by hardware limitations that were recently removed. So we can simplify.

57300 replied:
Updated. I changed my mind about the label rename, particularly because I wanted to keep `ipc_shm` labels as is.

57300 commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:
```diff
@@ -116,14 +116,6 @@
-	status = "okay";
-};
-
-&shared_ram3x_region {
```
`cpuapp_dma_region` is already enabled further down.

hubertmis replied:
Thanks, I missed it. It's not in the diff :)

57300 commented on include/zephyr/dt-bindings/reserved-memory/nordic-owned-memory.h:
```diff
@@ -0,0 +1,53 @@
+#define NRF_PERM_RW   (NRF_PERM_R | NRF_PERM_W)
+#define NRF_PERM_RX   (NRF_PERM_R | NRF_PERM_X)
+#define NRF_PERM_RS   (NRF_PERM_R | NRF_PERM_S)
+#define NRF_PERM_WX   (NRF_PERM_W | NRF_PERM_X)
```
Seems like it shouldn't be, unless it's technically allowed by the bus architecture. I'm not familiar enough with it.

Either way, it was also possible to set `perm-execute` without `perm-read` with the old binding, so I'm not sure we need to add new restrictions here.

hubertmis replied:
I'm pretty sure the bus allows for this, but I can't imagine a master that would send a transaction with an X attribute but without R.

My comment was intended to suggest removing the `define`s that won't be ever used. Just to avoid maintaining dead code. But I'm fine to keep it.

57300 committed:
boards: nordic: Flatten shared_ramxx_region nodes

Update the default memory maps for `nrf54h20dk` and `nrf9280pdk` to
remove the `shared_ram20_region` and `shared_ram3x_region` nodes,
because their child nodes no longer need to be grouped together:

  * IPC buffers in RAM20 are statically allocated.
  * DMA buffers in RAM3x have separate access owners.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
